### PR TITLE
Add Span.StartsWith(T) and EndsWith(T)

### LINKED ISSUE #87689: [API Proposal]: Span.StartsWith/EndsWith(T)
### Background and motivation

`string.StartsWith(char)` and `string.EndsWith(char)` are useful shortcuts for `this.Length > 0 && this[0] == 'a'`.
They also sometimes save you from having to declare temporary locals `SomeCall().StartsWith('a')`.
For strings, they are also more efficient now but that's not the main motivation and likely wouldn't apply to Span extensions.

The fact that Span is missing such a helper feels like an oversight on an otherwise helpful API surface of utility extensions and I often end up writing a helper class that polyfills these. A quick [GH search](https://github.com/search?q=%22StartsWith%28this+ReadOnlySpan%3Cchar%3E%22&type=code&ref=advsearch) shows ~10 other such places.

It also means that some analyzers light up on `string`, but not on `ReadOnlySpan<char>`, e.g. the recently added [CA1858](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1858) - "Use StartsWith instead of IndexOf".

I didn't find an existing issue for these so figured they're at least worth discussing.
At least unlike most? other extensions (e.g. `IndexOf`, `StartsWith(ReadOnlySpan, ReadOnlySpan)` ...), these methods are trivial to polyfill without losing out on potential perf.

Edit: more discussion around these APIs also at https://github.com/dotnet/runtime/discussions/85374

### API Proposal

```csharp
namespace System;

public static class MemoryExtensions
{
    // Existing
    public static bool StartsWith(this System.ReadOnlySpan<char> span, System.ReadOnlySpan<char> value, System.StringComparison comparisonType);
    public static bool StartsWith<T>(this System.ReadOnlySpan<T> span, System.ReadOnlySpan<T> value) where T : System.IEquatable<T>?;
    public static bool StartsWith<T>(this System.Span<T> span, System.ReadOnlySpan<T> value) where T : System.IEquatable<T>?;
    // ... same pattern for EndsWith

    // Proposed
    public static bool StartsWith<T>(this System.ReadOnlySpan<T> span, T value) where T : System.IEquatable<T>?;
    public static bool StartsWith<T>(this System.Span<T> span, T value) where T : System.IEquatable<T>?;
    public static bool EndsWith<T>(this System.ReadOnlySpan<T> span, T value) where T : System.IEquatable<T>?;
    public static bool EndsWith<T>(this System.Span<T> span, T value) where T : System.IEquatable<T>?;
}
```

### API Usage

```diff
-static bool IsNameHidden(ReadOnlySpan<char> fileName) => fileName.Length > 0 && fileName[0] == '.';
+static bool IsNameHidden(ReadOnlySpan<char> fileName) => fileName.StartsWith('.');
```

### Alternative Designs

While most helpers on `MemoryExtensions` are available for both `Span` and `ReadOnlySpan` of any `T`, the vast majority of use for this one will likely end up being `ReadOnlySpan<char>`.

```csharp
public static bool StartsWith(this System.ReadOnlySpan<char> span, char value);
public static bool EndsWith(this System.ReadOnlySpan<char> span, char value);
```

### Risks

_No response_

### PR DESCRIPTION
**Pull request by MihaZupan**
Closes #87689

### CODE REVIEW

File: src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs [hunk 1]
```diff
@@ -674,7 +674,7 @@ private static void BuildCommandLine(ProcessStartInfo startInfo, ref ValueString
             // problems (it specifies exactly which part of the string
             // is the file to execute).
             ReadOnlySpan<char> fileName = startInfo.FileName.AsSpan().Trim();
-            bool fileNameIsQuoted = fileName.Length > 0 && fileName[0] == '\"' && fileName[fileName.Length - 1] == '\"';
+            bool fileNameIsQuoted = fileName.StartsWith('"') && fileName.EndsWith('"');
```
Comment (stephentoub) [inline src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs:677]: Does the JIT successfully eliminate the double check on fileName.Length?

File: src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.cs [hunk 2]
```diff
@@ -2619,6 +2619,22 @@ public static unsafe int SequenceCompareTo<T>(this ReadOnlySpan<T> span, ReadOnl
...
+        public static bool StartsWith<T>(this ReadOnlySpan<T> span, T value) where T : IEquatable<T>? =>
+            span.Length != 0 && (span[0]?.Equals(value) ?? (object?)value is null);
+
+        /// <summary>
+        /// Determines whether the specified value appears at the end of the span.
+        /// </summary>
+        /// <param name="span">The span to search.</param>
+        /// <param name="value">The value to compare.</param>
+        public static bool EndsWith<T>(this ReadOnlySpan<T> span, T value) where T : IEquatable<T>? =>
+            span.Length != 0 && (span[^1]?.Equals(value) ?? (object?)value is null);
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.cs:2636]: Seems like the JIT might be having a little trouble optimizing this? [SharpLab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAMIR8AB14AbGFADKMgG68wMXAG4aNYgGYm5JE1IMhDAN40GlpruYHgECJIYBJXAEUOEDDAAm5ABRsMNg+APJckgCesmLYXAA8QgAW2FAAfAy4sVwAlAwAvGkWViVZcSwAolw+uADqvBhJAQDkAETNORrUJcWWOnp2Dk6uHl6+pIHBYRHR2Ykp6ZnZeYW9JUvlVTX1jRNtHV093Vb9tgz2ji7unt4+2pMh4VExcfOpGWW5BUXH6xtcLAAMjAuABzRoMACE+QYjAAZHD/gBtAB65AAupUAI4cbCSXD+fadNYk37WAbnIYMLZ1BpNeIAFTS/kavFwDCCjxmLwSTP+aAYDIYCjxHBgeQA7kkZDBBQwQC4KjjsBhsMBpIy0gB+b5rUrZIEg8FJKEw+GI/yfVEYrXY3H4/wiyRivJanX+CDAABWMDAGC1OSdYoYbIYXA4kkkxLJayRAFkYI0ID5nOJJP4E0mU2nQmI+BAuLgWABBUGg2C4XC8BQwZwRXhcBugnLotanfSUy40nZJUiallJUOc6bPOZ8z4CoVB8UMKUyuUK5xK3Gq9UwTU61Zkko+GAAM2wEYw/gZeVDXC8YYjTi1er+DEtBuBYIh0NhDARyLRmOXeIJ05yPIQDvP5PkNF8TTfc0Hytb9bV/B0AIYN0H09H0/QDacQ3ZcNI2jABfIA)

Comment (MihaZupan) [reply src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.cs:2636]: Should have improved with #102518 / #102808.
I can double-check what it's doing now.

@MihuBot -fast

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.cs (hunk 2 above)
Comment (MihaZupan) [reply src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.cs:2638]: It's generating the same ASM these now
```c#
[Benchmark]
public bool Old()
{
    ReadOnlySpan<char> span = Value;
    return span.Length != 0 && span[0] == '"' && span[^1] == '"';
}

[Benchmark]
public bool ExtGeneric()
{
    ReadOnlySpan<char> span = Value;
    return span.StartsWith('"') && span.EndsWith('"');
}

[Benchmark]
public bool ExtChar()
{
    ReadOnlySpan<char> span = Value;
    return span.StartsWithChar('"') && span.EndsWithChar('"');
}
```

File: src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs (hunk 1 above)
Comment (MihaZupan) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs:677]: Yes.
It does choose not to inline `EndsWith` if the input never has a leading `"` though, resulting in more code than if it did inline. Marked it as aggressive inlining now.

**Review by stephentoub — APPROVED** [review]

**Pull request merged**